libs/wingc/src/ast.rs

```diff
@@ -742,6 +742,8 @@ pub enum Reference {
 		property: Symbol,
 		optional_accessor: bool,
 	},
+	/// A reference to an accessed member of an object `expression[x]`
+	ElementAccess { object: Box<Expr>, index: Box<Expr> },
```

Comment:
I'd like to re-thing this as a "reference" and perhaps make `ElementAccess` just another `ExprKind`. References are are expressions that have a name but `ElementAccess` is just two expressions (object and index) with no name: `[1,2,3,4][0]` should evaluate to 1.

Reply:
The difference being that anything after the `.` is a name. Json is the exception with it's dotted access to non-fields, but for any typed "object" access you always access a name. Even in json there's no real way to access an expression after the `.`: `my_json.this_cant_be_an_expression`.

Reply:
> The difference being that anything after the . is a name.

True, but that's more of an optimization, both for our implementation and for readability. We could easily allow `a."b"` or even `a."{some string expression}"`. Not saying we should, but we could. 
The crux of my concern is that I'm not sure we're actually making use of idea that "References are expressions that have a name" anyways.

Again I'm actually in favor of breaking down `reference` into more useful parts. Like considering lvalues and rvalues for expressions and/or seeing `.` and `[]` as operators.

Reply:
Right, I wasn't thinking of the left side of the assignment. I recall explaining to someone in the past that in our AST "reference" is actually "lvalue", so I'm convinced what you did is the best option given the current design.
Maybe we should change `Reference` to `LValue`..